### PR TITLE
[explain] Support materialized views in `explain filter pushdown`

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

#24498

In particular, it can be helpful to see the statistics for an existing materialized view -- to know what the level of pushdown would be like if the view was re-bootstrapping now.

### Tips for reviewer

I think the diciest part of this PR is the timestamp selection code... please check my work!

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/sequencer/inner.rs [hunk 1]
```diff
@@ -1897,6 +1904,86 @@ impl Coordinator {
         };
     }
 
+    async fn render_explain_pushdown<'a>(
```
Comment (bkirwi) [inline src/adapter/src/coord/sequencer/inner.rs:1907]: This is extracted from `peek.rs` without meaningful changes.

File: src/adapter/src/coord/sequencer/inner/create_materialized_view.rs [hunk 2]
```diff
@@ -819,4 +819,50 @@ impl Coordinator {
...
+
+        let desc = self
+            .catalog
+            .try_get_physical_plan(&gid)
+            .ok_or_else(|| ExplainError::UnknownError(format!("unknown gid: {gid}")))?
+            .clone();
+
+        let id_bundle = dataflow_import_id_bundle(&desc, item.cluster_id);
+        let (_read_holds, dataflow_as_of, storage_as_of, until) =
+            self.select_timestamps(id_bundle, item.refresh_schedule.as_ref())?;
```
Comment (bkirwi) [inline src/adapter/src/coord/sequencer/inner/create_materialized_view.rs:843]: AIUI, this will:
- Get a set of frontiers for this MV as though it were being scheduled now. (Which is the desired behaviour for `explain filter pushdown`: we want to know what kind of pushdown we'd see if we were re-rehydrating the view right now.)
- Take read holds on the inputs, ensuring that the call to `render_...` below will be able to snapshot at the chosen storage as-of.

I think this seems right in the case of ordinary MVs. It's not clear to me what semantics we'd want in the case of `REFRESH EVERY` or similar, however! Certainly we don't want to sit around until the next refresh interval...

(If this has an obvious answer I'll toss it in; otherwise I may just ban it for now...)

**Review by jkosh44 — APPROVED** [review]
LGTM

**Review by aalexandrov — APPROVED** [review]
Minor nits, except what @ggevay [said here](https://github.com/MaterializeInc/materialize/pull/26467#r1556183487). I think that `bootstrap_dataflow_as_of` is the right way to implement this, because it is the top-level  method called by `Coordinator::bootstrap`.

File: src/adapter/src/coord/sequencer/inner/create_materialized_view.rs (hunk 2 above)
Comment (bkirwi) [reply src/adapter/src/coord/sequencer/inner/create_materialized_view.rs:843]: > so usually the timestamps returned by select_timestamps won't match the as_of that we would select if we were bootstrapping the MV now.

Interesting... do you have a good intuition for _why_ these methods behave so differently? (Is it just that we have the output MV and can make a less conservative choice?

I'm a bit hesitant to just invoke the bootstrap code, since it seems to run in time proportional to all collections, where I'm interested in just the one. But I'll experiment with that and the workaround you suggest!

Comment (ggevay) [reply src/adapter/src/coord/sequencer/inner/create_materialized_view.rs:843]: > do you have a good intuition for why these methods behave so differently?

Concentrating on the most common case (write frontier for bootstrap, and `least_valid_read` for new MV):
If the MV is already there, then we most commonly just need to continue where we left off writing it. But when we are creating the MV, then we have to make a decision about what timestamp it should start from. For the latter, `least_valid_read` is a natural choice, but `least_valid_read` is usually a second or so behind the write frontier, and can be arbitrarily behind in various corner cases, as mentioned above.

> (Is it just that we have the output MV and can make a less conservative choice?

Yes, in the above common case, we can say that `bootstrap_dataflow_as_of` is less conservative, because the dataflow will start at a later time.

But generally, I wouldn't say that it's less conservative, because there are also some constraints that `bootstrap_dataflow_as_of` needs to take into account that don't come up when creating a new MV:
- Compute reconciliation (i.e., reusing already-running dataflows when envd is restarted without clusterds being restarted). This is quite tricky, and is what necessitates looking at the `StorageConstraints` (which is the thing that takes time proportional to all collections).
- A REFRESH MV might be after its last REFRESH. (We don't allow this for newly created MVs.)

Comment (ggevay) [reply src/adapter/src/coord/sequencer/inner/create_materialized_view.rs:843]: > time proportional to all collections

If you are worried about this, then how about adding a histogram metric for how much time this takes? Then we could observe how it actually looks in practice. And then if it turns out to be actually too slow, then we'll figure something out. (For example, we could write a version of `collect_dataflow_storage_constraints` that looks at not all collections but only at the ones relevant to one EXPLAIN command.)

Btw. there are things that we do every second that take time proportional to the number of collections:
- `advance_timelines` calls the rather complicated `update_read_holds` -> `set_read_policy` -> `update_read_capabilities` for every index and matview, and also iterates through all catalog objects.
- `ActiveInstance::update_write_frontiers` -> `update_read_capabilities` does really complex stuff for every Compute collection.

Comment (bkirwi) [reply src/adapter/src/coord/sequencer/inner/create_materialized_view.rs]: Cool - if none of you are worried about blocking the coordinator thread, then neither am I! Updated the implementation to use `bootstrap_dataflow_as_of`, imitating how it's called when bootstrapping an MV.

**Comment by bkirwi** [comment]
Okay, I have attempted to apply all suggestions - should be ready for another look.

Output looks reasonable at a glance:

```
materialize=> create materialized view check as select * from bids where bid_time + '15 seconds' > mz_now();
CREATE MATERIALIZED VIEW

materialize=> explain filter pushdown for materialized view mv;
         Source          | Total Bytes | Selected Bytes | Total Parts | Selected Parts 
-------------------------+-------------+----------------+-------------+----------------
 materialize.public.bids | 27643       | 11342          | 15          | 10
(1 row)

materialize=> explain filter pushdown for materialized view mv;
         Source          | Total Bytes | Selected Bytes | Total Parts | Selected Parts 
-------------------------+-------------+----------------+-------------+----------------
 materialize.public.bids | 22943       | 6642           | 10          | 5
(1 row)
```

We see the number of bytes / parts that need to be fetched vary, but remain roughly constant even as the shard grows.

File: test/sqllogictest/explain-pushdown.slt [hunk 3]
```diff
@@ -95,6 +95,16 @@ EXPLAIN FILTER PUSHDOWN FOR SELECT * FROM jsonb_fields where timestamp > 1000 AN
...
+# EXPLAIN FILTER PUSHDOWN FOR MATERIALIZED VIEW is also supported
+
+statement ok
+CREATE MATERIALIZED VIEW big_numbers AS SELECT * FROM numbers WHERE value > 10000;
+
+query TIIII
+EXPLAIN FILTER PUSHDOWN FOR MATERIALIZED VIEW big_numbers
+----
+materialize.public.numbers  0  0  0  0
+
```
Comment (ggevay) [inline test/sqllogictest/explain-pushdown.slt:107]: You could maybe add some more tests. For example, you could call it on all the MVs in `materialized_views.slt`. You can get a list of MV names in that file by looking at the output of the second to last introspection query. If it's hard to check the concrete output you could just do `statement ok` on them.

**Comment by ggevay** [comment]
Thanks, the new `as_of` selection looks good! But some more tests would be good to have, I think.

File: test/sqllogictest/explain-pushdown.slt (hunk 3 above)
Comment (bkirwi) [reply test/sqllogictest/explain-pushdown.slt:107]: Yeah, the output for (non-empty) materialized views is not necessarily stable, since they reveal the exact part splits / byte counts which may change. But I can do `statement ok`!

File: src/adapter/src/coord/sequencer/inner.rs [hunk 4]
```diff
@@ -2079,6 +2086,92 @@ impl Coordinator {
...
+        import_as_of: Antichain<Timestamp>,
+        imports: impl IntoIterator<Item = (GlobalId, MapFilterProject)>,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        use futures::TryStreamExt;
+
+        if import_as_of.is_empty() {
+            Err(AdapterError::Internal(
+                "cannot explain stats as-of the empty antichain".to_string(),
+            ))?;
+        }
```
Comment (bkirwi) [inline src/adapter/src/coord/sequencer/inner.rs:2102]: I've added this clause because the added SLTs showed that we were trying to fetch stats for inputs that had advanced to `[]`, which can happen with constant MVs and similar.

Right now I've made this an error, but it would also be possible to make this a success with zero rows... under the theory that in practice this will render no sources at all next time the cluster rehydrates. Either seems fine to me, but I'm happy to swap this if any readers have a preference.

Comment (ggevay) [reply src/adapter/src/coord/sequencer/inner.rs:2102]: I'd say erroring is ok: If a user is calling EXPLAIN FILTER PUSHDOWN on an MV that won't create a dataflow on the next restart (either because it's a constant, or it's past its last refresh, etc.), that feels like some kind of mistake, so it's good if it's called out.

But maybe make it a non-internal error. I think `AdapterError::Internal` is reserved for situations where there is some bug in the system.

**Comment by bkirwi** [comment]
Taking back to draft as I integrate some of the recent explain changes and try and work through some corner cases in the bootstrap logic...

**Comment by bkirwi** [comment]
After switching to `bootstrap_dataflow_as_of` and adding tests for the weird menagerie of MVs in `materialized_views.slt`, I'm now running into [this panic](https://buildkite.com/materialize/test/builds/83096#018fdfc7-e099-4066-ab90-3e1418960716).

```
thread 'coordinator' panicked at src/adapter/src/coord.rs:2339:13:
error bootstrapping dataflow `as_of`: `min_as_of` [1717446599000] greater than `max_as_of` [1717446595426] (import_ids=[u57], export_ids=[u59], storage_constraints=StorageConstraints { dependencies: {User(57)}, dependants: {User(59)} })
stack backtrace:
0: rust_begin_unwind
1: core::panicking::panic_fmt
2: <mz_adapter::coord::Coordinator>::bootstrap_dataflow_as_of
```

`min_as_of` is chosen based on the read holds we're able to obtain for all inputs. `max_as_of` is based on the meet of the `since.join(upper)` of all outputs. (Colloquially, that's the time that we need to make sure we're generating correct output as of... it's the least frontier such that future writes at that time will be visible to readers.)

This assert seems intuitively correct, though I'm not 100% sure that it's valid in the case of refresh-every MVs. Since it's only showing up when this method is invoked on my branch, that implies that there's something that holds true during bootstrap but not afterwards, but I haven't figured out which.

I'm going to try and bisect this, but given that it's just a flake, that might take a long time or give wrong results. If anyone has ideas or suggestions for places to look, I'm all eyes!

**Comment by ggevay** [comment]
I'll take a look tomorrow!

**Comment by bkirwi** [comment]
Thanks! If it's useful, the error seems to reliably be when explaining the materialized view `mvi1`, which is defined as:

```
CREATE MATERIALIZED VIEW mvi1 WITH (REFRESH EVERY '8s' ALIGNED TO mz_now()::string::int8 + 2000) AS
SELECT 5*x FROM t3;
```

It is indeed a refresh-every view, though I'm not sure what in particular would make it the troublesome ones. I'm seeing a failure once in every ~5 runs of the SLT on average. The `min_as_of` appears reliably just a few seconds past the `max_as_of`.

**Comment by ggevay** [comment]
> This assert seems intuitively correct, though I'm not 100% sure that it's valid in the case of refresh-every MVs.

I don't see any reason why this shouldn't be true for REFRESH MVs.

> Since it's only showing up when this method is invoked on my branch, that implies that there's something that holds true during bootstrap but not afterwards, but I haven't figured out which.

Well, not necessarily. It's also possible that our restart tests are somehow not hitting the same issue due to just random chance. In other words, maybe if we did a restart at the exact same moment where your EXPLAIN FILTER PUSHDOWN is running into the issue, then maybe bootstrapping would run into the same issue too.

I've looked into this, but it appears to be some deep bug. `bootstrap_dataflow_as_of` seems correct: The assertion must be true simply because the Compute Controller ensures that for every materialized view, we have a read capability on every input at a time that is not greater than the write frontier of the materialized view. So there might be some bug in the controllers' frontier management:
1. Maybe the Compute Controller is somehow not keeping a correct read capability on an input.
2. Or maybe the Storage Controller somehow forgets about a read capability that the Compute Controller wants to keep.
3. Or maybe the Storage Controller somehow gets confused about what is the write frontier of the MV. Like, the MV writes some output data, so a new write frontier gets to `handle_frontiers_response` and the Compute Controller updates the input read holds, but somehow the Storage Controller doesn't update the write frontier for the MV. Btw. where is the code that updates the Storage Controller's knowledge of an MV's write frontier whenever the MV writes some output? Wherever that is, might it be that the information gets there later than to the Compute Controller's `handle_frontiers_response`?

I'm leaning towards 3., because I see no reason for why would the Compute Controller's input read holds even matter in this situation: the MV is a very simple computation, reading from a single table, so as soon as the table's write frontier ticks forward, the MV should be able to immediately process it, and therefore there should be no reason to keep the table's read capability from advancing for several seconds (the difference in the assert msg that Ben pasted is more than 3 seconds).

@teskje, what do you think?

> I'm seeing a failure once in every ~5 runs of the SLT on average.

It immediately happened for me locally on the first try, but then I added some debug logging, and it hasn't happened since, even though I've ran it much more than 5 times by now :D

**Comment by ggevay** [comment]
> Btw. where is the code that updates the Storage Controller's knowledge of an MV's write frontier whenever the MV writes some output? Wherever that is, might it be that the information gets there later than to the Compute Controller's `handle_frontiers_response`?

This might be it! I've added an assert in `handle_frontiers_response` that queries the Storage Controller for the MV's write frontier, and checks that it's not earlier than the frontier that `handle_frontiers_response` was called with, and this assert fires super quickly in the slt! My debug branch is [here](https://github.com/ggevay/materialize/tree/bootstrap-datalfow-as_of-panic). (I have to sign off for today.)

**Comment by ggevay** [comment]
Btw. if indeed the above is the problem, then the actual bootstrapping might be safe from it, because the Storage Controller's state is rebuilt during restart. (Assuming that the updated write frontier hits Persist before it reaches `handle_frontiers_response`.)

**Comment by teskje** [comment]
> Btw. where is the code that updates the Storage Controller's knowledge of an MV's write frontier whenever the MV writes some output? Wherever that is, might it be that the information gets there later than to the Compute Controller's handle_frontiers_response?

Yep, that’s possible now. Until recently the compute controller would tell the storage controller about write frontier updates of MVs, but since https://github.com/MaterializeInc/materialize/pull/27057 the `StorageCollections` object installs a background task that listens for these frontier advancements on its own using persist handles. I would expect that to usually be faster than going through the compute protocol, but that might be different in CI. In any case, the storage controller’s view of MV write frontiers can be different than the compute controller’s.

**Comment by bkirwi** [comment]
Yeah, that all sounds quite plausible! In particular it helps explain why we don't see issues at bootstrap time. (Though of course this appears pretty hard to trigger even in the problem case.)

I suppose possible fixes include:
1. Having the compute controller learn about progress via storage collections directly. (I like this but it seems like a significant change.)
2. Making the frontier-fetch method do a linearizable upper fetch. (Definitely fine but maybe expensive.)
3. Turning the panics into errors and adding a retry loop. (Gross but probably works around this already-esoteric issue.)

I'll try whichever one of these is easiest and follow up with whether I can still reproduce or not, to help confirm the diagnosis.

**Comment by ggevay** [comment]
Yeah, I agree that we should consider moving away from `bootstrap_dataflow_as_of`. At this point, it's probably better to not exactly nail the timestamp that an MV would have if a restart were to happen, rather than to introduce more code complications.

**Comment by ggevay** [comment]
> > Would it work to just use the current read frontier of the MV dataflow, i.e. the read frontier the compute controller knows for that collection?

> IIUC the current read frontier of an MV is no later than the write frontier, and we're getting this panic because that's earlier than any of the read holds we can obtain. So I feel like the direct version of this might panic as well.

I think Jan was suggesting to get the read frontier of the MV dataflow from the Compute Controller, which should be consistent with the read holds that the Compute Controller is keeping on the inputs. So you should be able to read the inputs at that timestamp.

**Comment by teskje** [comment]
> Which makes sense to me - it's the ultimate source of truth for what rehydrating a dataflow would do, IIUC!

Depends on the circumstances of rehydrating: `bootstrap_dataflow_as_of` selects the as-of from which dataflows hydrate after an envd restart. But if you just spawn a new replica in a cluster, it will hydrate the dataflow from its current read frontier.

> IIUC the current read frontier of an MV is no later than the write frontier, and we're getting this panic because that's earlier than any of the read holds we can obtain.

This might be true. I think you can only get a read hold at a storage collection's `implied_capability`, where the read frontier of a dataflow can easily be before that. The read frontier can never be before the `read_capabilities.frontier()` of any of its input collections, but the storage controller might refuse to give out read holds for that time.

I don't think there are any hard reasons for the storage controller to behave like that, so that might be something we can change. But I agree, using `max(read_hold, compute_read_frontier)` seems fine.

**Comment by bkirwi** [comment]
Cool! I'm not totally comfortable with my understanding of the new offset-hold infrastructure, but the approach does seem to work... see the final commit for the relevant diff.

Nightly run [here](https://buildkite.com/materialize/nightly/builds/8014), though it's probably not that interesting.

File: test/sqllogictest/materialized_views.slt [hunk 5]
```diff
@@ -1619,3 +1619,118 @@ create  cluster-replica  "c_schedule_1"  "schedule"  false  {"decision":"on","ob
...
+statement error db error: ERROR: canceling statement due to statement timeout
+EXPLAIN FILTER PUSHDOWN FOR MATERIALIZED VIEW mv10;
+
+statement error db error: ERROR: canceling statement due to statement timeout
+EXPLAIN FILTER PUSHDOWN FOR MATERIALIZED VIEW mv11;
+
+simple conn=mz_system,user=mz_system
+ALTER SYSTEM SET enable_explain_pushdown = false
+----
+COMPLETE 0
```
Comment (ggevay) [inline test/sqllogictest/materialized_views.slt:1736]: Maybe you could also set the `statement_timeout` back to its default value. (Although, I think sqllogictest mostly resets the system between slt files, but still...)

File: test/sqllogictest/materialized_views.slt [hunk 6]
```diff
@@ -1619,3 +1619,118 @@ create  cluster-replica  "c_schedule_1"  "schedule"  false  {"decision":"on","ob
...
+simple conn=mz_system,user=mz_system
+ALTER SYSTEM SET enable_explain_pushdown = true
+----
+COMPLETE 0
+
+# Pulling stats for refresh-every and similar MVs can time out,
+# since data may not yet be available...
+
+simple conn=mz_system,user=mz_system
+ALTER SYSTEM SET statement_timeout = '1s'
```
Comment (ggevay) [inline test/sqllogictest/materialized_views.slt:1634]: This seems to be not taking effect. If you run the slt with `-vv`, which prints every statement, it's visible that the last 3 statements each take about 10 seconds, which is the default statement timeout. Maybe this `ALTER` doesn't affect existing sessions? There is a notice that kind of suggests this:
```
materialize=> ALTER SYSTEM SET statement_timeout = 1;
NOTICE:  variable "statement_timeout" was updated for the system, this will have no effect on the current session
ALTER SYSTEM
```
Although this talks about the current session, maybe it's also true for any other ongoing session?

@MaterializeInc/testing , is there a way to reliably change the statement timeout in an slt?

Also, should we change the message of this notice if the reality is that indeed any other ongoing session is not affected?

(Alternatively, you could just not bother with the statement timeout, but explain only one of these MVs where the explain will block.)

**Review by ggevay — CHANGES_REQUESTED** [review]

**Comment by ggevay** [comment]
The timestamp selection is looking good to me, thanks for iterating on this!

I wrote some comments on the slt.

File: test/sqllogictest/materialized_views.slt (hunk 6 above)
Comment (nrainer-materialize) [reply test/sqllogictest/materialized_views.slt:1634]: Not sure, could maybe testdrive's `$ set-sql-timeout` help you?

Comment (ggevay) [reply test/sqllogictest/materialized_views.slt:1634]: I guess that could work. So, instead of this slt, you could test blocking explains in `materialized-view-refresh-options.td`, where e.g. `mv9` and `mv10` should also be non-readable, similarly to these MVs.

Comment (bkirwi) [reply test/sqllogictest/materialized_views.slt]: Ah! This should be `SET statement_timeout = '1s'`, not `ALTER SYSTEM SET statement_timeout = 1;`, if we want it to just apply to the current session.

I've pushed a commit with the proper syntax. I think this addresses your first comment too... since it's session local there's no need to unset it.

> If you run the slt with `-vv`

I was not aware of this option... thanks for mentioning! Will save me some time in the future.

**Review by ggevay — APPROVED** [review]
